# Data Preparation

`Overview`
This notebook handles the initial data processing pipeline:
- Loading raw data from source files
- Performing exploratory data analysis (EDA)
- Cleaning and handling missing values
- Feature preprocessing and engineering
- Exporting processed datasets for modeling

`Inputs`
- Raw data files from `../data/raw/` 

`Outputs`
- Processed datasets in `../data/processed/`
- EDA visualizations in `../reports/figures/`

`Dependencies`
- pandas
- numpy
- matplotlib
- seaborn

*Note: This is notebook 1 of the analysis pipeline*

In [1]:
# Imports 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
from pathlib import Path

# Import custom modules
# from src.save_load import save_parquet

In [2]:
!which python


/Users/bertdepoorter/opt/anaconda3/envs/MDA/bin/python


Here we load the project specific datasets as CSV files. In the follow-up cell, we load the auxiliary dataset containing extra information on the CORDIS-HORIZON projects. This includes
- Scientific vocabulary 
- legal basis documents
- organization
- project
- topics
- webItem 
- webLink

In [5]:
# Import the dataset as pandas DataFrame
run_dir = os.getcwd()
parent_dir = os.path.dirname(run_dir)

raw_dir = f'{parent_dir}/data/raw'
interim_dir = f'{parent_dir}/data/interim'
processed_dir = f'{parent_dir}/data/processed'

# define file paths to project-specific files
data_report_path = f'{raw_dir}/reportSummaries.csv'
data_filereport_path = f'{raw_dir}/file_report.csv'
data_publications_path = f'{raw_dir}/projectPublications.csv'
data_deliverables_path = f'{raw_dir}/projectDeliverables.csv'



## Inspect Reports

In [6]:
# get DataFrame keys
data_report = pd.read_csv(data_report_path, delimiter=';')
data_report.keys()

Index(['id', 'title', 'projectID', 'projectAcronym', 'attachment',
       'contentUpdateDate', 'rcn'],
      dtype='object')

In [8]:
data_report.head()

id                                              title  \
0  101066069_PSHORIZON  Periodic Reporting for period 1 - ERASMUS (Ear...   
1  101073231_PSHORIZON  Periodic Reporting for period 1 - OncoProTools...   
2  101068156_PSHORIZON  Periodic Reporting for period 1 - BLISS (Beta-...   
3  101072180_PSHORIZON  Periodic Reporting for period 1 - Green2Ice (W...   
4  101063407_PSHORIZON  Periodic Reporting for period 1 - GHost (His E...   

   projectID projectAcronym  \
0  101066069        ERASMUS   
1  101073231   OncoProTools   
2  101068156          BLISS   
3  101072180      Green2Ice   
4  101063407          GHost   

                                          attachment    contentUpdateDate  \
0                                                NaN  2025-03-17 10:38:00   
1  /docs/results/horizon/101073/101073231_PS/2024...  2025-03-18 12:31:34   
2  /docs/results/horizon/101068/101068156_PS/pict...  2025-03-05 11:47:45   
3  /docs/results/horizon/101072/101072180_PS/2023...  2025-02-14 10:36:27   
4  /docs/results/horizon/101063/101063407_PS/pa-1...  2025-02-26 17:32:14   

       rcn  
0  1267558  
1  1270628  
2  1260626  
3  1252991  
4  1257475

### Missing values
1. we check each column for missing values
2. Define decision tree for handling missing values
3. Change values algorithmically
4. Store update dataframe in interim directory


In [9]:
# look for missing values
report_missing = data_report.isnull()

# check which columns are missing data
for key in data_report:
    missing = report_missing[report_missing[key] == True]
    print(f'For key {key}:\n     {len(missing.id)} elements are missing.')

For key id:
     0 elements are missing.
For key title:
     0 elements are missing.
For key projectID:
     0 elements are missing.
For key projectAcronym:
     0 elements are missing.
For key attachment:
     1861 elements are missing.
For key contentUpdateDate:
     0 elements are missing.
For key rcn:
     0 elements are missing.


We see that there are only missing attachments. These attachments refer to some additional documents, mostly png picture.
We can handle this in three ways
- Look manually for the missing attachments 
- Ignore this column during analysis
- if attachment present: add to dashboard when user wants to inspect a particular project. If not present: leave blank. 

I recommend using the last approach. 

In [10]:
# handle missing values

# define missng values rule here

# change the missing values in dataframe
project_reports_interim = data_report
# save updated dataframe to data/interim
project_reports_interim.to_csv(f'{interim_dir}/reportSummaries_interim.csv', sep=';')

### Inspect other report file
This CSV file does not contain useful information

In [11]:
data_filereport = pd.read_csv(data_filereport_path, delimiter=';')
data_filereport

filename,status, issue_cause downloadURL, issue_cause accessURL
0   HORIZON Report summaries (individual XML files...             
1                        HORIZON Projects,delivered,,             
2           HORIZON Projects Deliverables,delivered,,             
3   HORIZON Projects (individual XML files),delive...             
4   HORIZON Projects Deliverables (individual XML ...             
5                HORIZON Report summaries,delivered,,             
6                    HORIZON Publications,delivered,,             
7           HORIZON Projects Deliverables,delivered,,             
8                    HORIZON Publications,delivered,,             
9           HORIZON Projects Deliverables,delivered,,             
10  HORIZON Publications (individual XML files),is...             
11               HORIZON Report summaries,delivered,,             
12                   HORIZON Publications,delivered,,             
13                       HORIZON Projects,delivered,,             
14               HORIZON Report summaries,delivered,,             
15                       HORIZON Projects,delivered,,             
16  CORDIS HORIZON organisations' collaboration ne...

## Inspect deliverables

In [12]:
# Inspect Dataframe
data_deliverables = pd.read_csv(data_deliverables_path, delimiter=';')
data_deliverables.keys()

Index(['id', 'title', 'deliverableType', 'description', 'projectID',
       'projectAcronym', 'url', 'collection', 'contentUpdateDate', 'rcn'],
      dtype='object')

In [13]:
data_deliverables

id  \
0      101071179_10_DELIVHORIZON   
1       101072491_9_DELIVHORIZON   
2       101066116_3_DELIVHORIZON   
3      101064988_11_DELIVHORIZON   
4      101071179_51_DELIVHORIZON   
...                          ...   
21919  101091749_16_DELIVHORIZON   
21920  101091749_30_DELIVHORIZON   
21921  101091749_29_DELIVHORIZON   
21922  101091749_33_DELIVHORIZON   
21923  101091749_25_DELIVHORIZON   

                                                   title  \
0        Technical/scientific review meeting 2 documents   
1                                MIRELAI project website   
2           Communication, Dissemination & Outreach Plan   
3      SINFONICA Knowledge map creation and System Ar...   
4                      Report on Portfolio activities 01   
...                                                  ...   
21919                        Report on DBL state of play   
21920                                  Demo-BLog website   
21921                         Demo-BLog project identity   
21922                            Promotional films (No1)   
21923  First release: Dissemination, Communication an...   

                              deliverableType  \
0                          Documents, reports   
1      Websites, patent fillings, videos etc.   
2                          Documents, reports   
3                          Documents, reports   
4                          Documents, reports   
...                                       ...   
21919                      Documents, reports   
21920  Websites, patent fillings, videos etc.   
21921                                   Other   
21922  Websites, patent fillings, videos etc.   
21923                      Documents, reports   

                                             description  projectID  \
0                         Draft agenda and presentations  101071179   
1                                MIRELAI project website  101072491   
2      The plan describes the planned measures to max...  101066116   
3      The deliverable will be the output of Tasks 4....  101064988   
4      The report will present the collaboration acti...  101071179   
...                                                  ...        ...   
21919                        Report on DBL state of play  101091749   
21920                                  Demo-BLog website  101091749   
21921                         Demo-BLog project identity  101091749   
21922                            Promotional films (No1)  101091749   
21923  First release: Dissemination, Communication an...  101091749   

      projectAcronym                                                url  \
0            SUSTAIN  https://ec.europa.eu/research/participants/doc...   
1            MIRELAI  https://ec.europa.eu/research/participants/doc...   
2        Ship Clones  https://ec.europa.eu/research/participants/doc...   
3          SINFONICA  https://ec.europa.eu/research/participants/doc...   
4            SUSTAIN  https://ec.europa.eu/research/participants/doc...   
...              ...                                                ...   
21919      Demo-BLog  https://ec.europa.eu/research/participants/doc...   
21920      Demo-BLog  https://ec.europa.eu/research/participants/doc...   
21921      Demo-BLog  https://ec.europa.eu/research/participants/doc...   
21922      Demo-BLog  https://ec.europa.eu/research/participants/doc...   
21923      Demo-BLog  https://ec.europa.eu/research/participants/doc...   

                collection    contentUpdateDate        rcn  
0      Project deliverable  2025-02-03 17:40:48  1246973.0  
1      Project deliverable  2025-02-26 08:02:37  1256521.0  
2      Project deliverable  2025-02-03 17:40:42  1246926.0  
3      Project deliverable  2025-02-03 17:40:44  1246937.0  
4      Project deliverable  2025-02-03 17:40:44  1246934.0  
...                    ...                  ...        ...  
21919  Project deliverable  2024-05-10 13:03:27  1092491.0  
21920  Project deliverable  2024-05-10 13:03:

I have changed the following lines to enable opening the file with pandas:
- 1412: something wrong in the deliverable description
- 1412: wrong use of delimiter
- 6677: wrong use of quotation marks
- 6678: wrong use of quotation marks
- 8812: use of delimiter inside string
- 8826: use of delimiter inside string
- 9360: use of delimiter inside string
- 9524: use of delimiter inside string
- 10128: use of delimiter inside string
- 13108: use of delimiter inside string
- 19931: use of delimiter inside string

### Missing values
Here we handle the missing values in the dataset

In [14]:
# look for missing values
deliverables_missing = data_deliverables.isnull()

# check which columns are missing data
for key in deliverables_missing.keys():
    missing = deliverables_missing[deliverables_missing[key] == True]
    print(f'For key {key}:\n     {len(missing.id)} elements are missing.')

For key id:
     0 elements are missing.
For key title:
     0 elements are missing.
For key deliverableType:
     35 elements are missing.
For key description:
     15 elements are missing.
For key projectID:
     0 elements are missing.
For key projectAcronym:
     0 elements are missing.
For key url:
     1 elements are missing.
For key collection:
     0 elements are missing.
For key contentUpdateDate:
     0 elements are missing.
For key rcn:
     1 elements are missing.


We are missing elements in the following columns:
- deliverableType
    - option 1: change to `'other'`
    - option 2: look up individual titles and add manually
- description
    - option 1: add empty string
    - Inspect manually to gain more insight what they exactly represent
        - Update: all the titles related to the projects are quite related. I suggest we copy title values into the description column.
- url
    - 1 missing url. Add the url to the main page of this project (SELFY, id = 101069748_16_DELIVHORIZON) instead of link to deliverable?
- rcn
    - 1 rcn is missing. 
    - Looked this number up in publication list based on the projectAcronym = `'GeneBEcon'`. There the rcn number is gives as `1077637.0`


In [15]:
# change unknown deliverable types to other
data_deliverables['deliverableType'] = data_deliverables['deliverableType'].fillna('Other') 

# change empty descriptions to title of that particular row
data_deliverables['description'] = data_deliverables['description'].fillna(data_deliverables['title'])

# change missing url to homepage of the particular project
data_deliverables['url'] = data_deliverables['url'].fillna('https://selfy-project.eu/')

# add missing rcn number
data_deliverables['rcn'] = data_deliverables['rcn'].fillna(1077637.0)

In [16]:
# check whether filling executed correctly
data_deliverables[deliverables_missing.deliverableType == True]

id                 title deliverableType  \
823    101062643_1_DELIVHORIZON  Data Management Plan           Other   
963    101039920_1_DELIVHORIZON  Data Management Plan           Other   
973    101040355_1_DELIVHORIZON  Data Management Plan           Other   
2526   101066209_1_DELIVHORIZON  Data Management Plan           Other   
4173   101068487_1_DELIVHORIZON  Data Management Plan           Other   
4174   101062409_1_DELIVHORIZON  Data Management Plan           Other   
4179   101060143_1_DELIVHORIZON  Data Management Plan           Other   
4180   101063914_1_DELIVHORIZON  Data Management Plan           Other   
5733   101043589_1_DELIVHORIZON  Data Management Plan           Other   
5740   101052538_1_DELIVHORIZON  Data Management Plan           Other   
5756   101043998_1_DELIVHORIZON  Data Management Plan           Other   
6247   101067706_1_DELIVHORIZON  Data Management Plan           Other   
7025   101090282_1_DELIVHORIZON  Data Management Plan           Other   
7027   101054860_1_DELIVHORIZON  Data Management Plan           Other   
7030   101039436_1_DELIVHORIZON  Data Management plan           Other   
8382   101065303_1_DELIVHORIZON  Data Management Plan           Other   
8383   101061526_1_DELIVHORIZON  Data Management Plan           Other   
9146   101064614_1_DELIVHORIZON  Data Management Plan           Other   
9152   101062923_1_DELIVHORIZON  Data Management Plan           Other   
9808   101067960_1_DELIVHORIZON  Data Management Plan           Other   
11214  101044612_1_DELIVHORIZON                   DMP           Other   
12656  101062365_1_DELIVHORIZON  Data Management Plan           Other   
14353  101067835_1_DELIVHORIZON  Data Management Plan           Other   
14444  101062241_1_DELIVHORIZON  Data Management Plan           Other   
14599  101065356_1_DELIVHORIZON  Data Management Plan           Other   
14610  101063066_1_DELIVHORIZON  Data Management Plan           Other   
14611  101060616_1_DELIVHORIZON  Data management plan           Other   
14916  101068449_1_DELIVHORIZON   Data Managment Plan           Other   
14943  101060987_1_DELIVHORIZON  Data Management Plan           Other   
14988  101069104_1_DELIVHORIZON  Data Management Plan           Other   
15343  101062988_1_DELIVHORIZON  Data Management Plan           Other   
17035  101070800_4_DELIVHORIZON  Data Management Plan           Other   
17558  101070847_2_DELIVHORIZON  Data Management Plan           Other   
17617  101057156_1_DELIVHORIZON  Data Management Plan           Other   
21439  101068805_1_DELIVHORIZON  Data Management Plan           Other   

                                             description  projectID  \
823    The Data Management Plan describes the data ma...  101062643   
963                                 Data Management Plan  101039920   
973                     Drafting of Data Management Plan  101040355   
2526   The Data Management Plan describes the data ma...  101066209   
4173   The Data Management Plan describes the data ma...  101068487   
4174   The Data Management Plan describes the data ma...  101062409   
4179   The Data Management Plan describes the data ma...  101060143   
4180   The Data Management Plan describes the data ma...  101063914   
5733                                Data Management Plan  101043589   
5740                                Data Management Plan  101052538   
5756                         Drafting of Data management  101043998   
6247   The Data Management Plan describes the data ma...  101067706   
7025   The Data Management Plan describes the data ma...  101090282   
7027                    Creation of Data Management Plan  101054860   
7030                    Drafting of Data Management plan  101039436   
8382   The Data Management Plan describes the data ma...  101065303   
8383   The Data Management Plan describes the data ma...  101061526   
9146   The Data Management Plan describes the data ma...  101064614   
9152   The Data Management Plan describes the data 

In [17]:
# save updated dataframe to data/interim
data_deliverables.to_csv(f'{interim_dir}/projectdeliverables_interim.csv', sep=';')

## Inspect Publications

In [18]:
# Inspect Dataframe
data_publications = pd.read_csv(data_publications_path, delimiter=';')
data_publications.keys()

Index(['id', 'title', 'isPublishedAs', 'authors', 'journalTitle',
       'journalNumber', 'publishedYear', 'publishedPages', 'issn', 'isbn',
       'doi', 'projectID', 'projectAcronym', 'collection', 'contentUpdateDate',
       'rcn'],
      dtype='object')

Some entries in the publications CSV have been changed by hand, in order to allow loading them:
- 7588
- 7748

Both are from the same conference. Problem: switched the order of the columns and add one additional empty column causing pandas loader to crash. 

Next problems:
- 12036: wrong notation of authors names + use of ; delimiter inside string.
- 12043: start authors string with four " + use ; to separate names.
- 12099: same problem as stated above
- 12110: same problem
- 12115: same problem
- 12270: same problem
- 18735: same problem
- 24019: same problem




In [19]:
data_publications

id  \
0      101040480_113381_PUBLIHORIZON   
1      101040480_113371_PUBLIHORIZON   
2      101040480_113383_PUBLIHORIZON   
3      101040480_113375_PUBLIHORIZON   
4      101040480_113374_PUBLIHORIZON   
...                              ...   
24145   101060809_11542_PUBLIHORIZON   
24146   101061890_16827_PUBLIHORIZON   
24147   101061201_13995_PUBLIHORIZON   
24148   101061201_13826_PUBLIHORIZON   
24149   101060809_11541_PUBLIHORIZON   

                                                   title  \
0      The Microwave Rotational Electric Resonance (R...   
1      The microwave spectra of the conformers of n-b...   
2      Coupled internal rotations and 14N quadrupole ...   
3      The Heavy Atom Structure, <i>“cis</i> effect” ...   
4      Structure determination of 2,5-difluorophenol ...   
...                                                  ...   
24145                    Review of International Studies   
24146      Dual-band electro-optically steerable antenna   
24147  Positron Annihilation Study of RPV Steels Radi...   
24148     Round Robin Tests for WWER Heat Exchange Tubes   
24149        European Journal of International Relations   

                isPublishedAs  \
0      Peer reviewed articles   
1      Peer reviewed articles   
2      Peer reviewed articles   
3      Peer reviewed articles   
4      Peer reviewed articles   
...                       ...   
24145  Peer reviewed articles   
24146  Peer reviewed articles   
24147  Peer reviewed articles   
24148  Peer reviewed articles   
24149  Peer reviewed articles   

                                                 authors  \
0      Hamza El Hadki, Kenneth J. Koziol, Oum Keltoum...   
1      Susanna L. Stephens, Eléonore Antonelli, Alexa...   
2       Mike Barth, Isabelle Kleiner, Ha Vinh Lam Nguyen   
3      Truong Anh Nguyen, Isabelle Kleiner, Martin Sc...   
4      K.P. Rajappan Nair, Kevin G. Lengsfeld, Philip...   
...                                                  ...   
24145               Pogdda, S., Richmond, O., Visoka, G.   
24146  Dmytro Vovchuk, Anna Mikhailovskaya, Dmitry Do...   
24147  Vladimir Slugen; Tomas Brodziansky; Jana Simeg...   
24148  Roman Krajcovic, Michal Benak, Radim Kopriva, ...   
24149   Oliver P. Richmond; Sandra Pogodda; Gëzim Visoka   

                                      journalTitle journalNumber  \
0                                        Molecules            28   
1                Journal of Molecular Spectroscopy           397   
2                  The Journal of Chemical Physics           160   
3                                     ChemPhysChem            25   
4                   Journal of Molecular Structure          1321   
...                                            ...           ...   
24145              Review of International Studies             3   
24146                            Journal of Optics     25 105601   
24147  Materials; Volume 15; Issue 20; Pages: 7091           NaN   
24148    e-Journal of Nondestructive Testing 28(7)           NaN   
24149  European Journal of International Relations           NaN   

       publishedYear publishedPages       issn isbn  \
0               2023            NaN  1420-3049  NaN   
1               2024            NaN  0022-2852  NaN   
2               2024            NaN  0021-9606  NaN   
3               2024            NaN  1439-4235  NaN   
4               2024            NaN  0022-2860  NaN   
...              ...            ...        ...  ...   
24145           2023            NaN  0260-2105  NaN   
24146           2023            NaN  2040-8986  NaN   
24147           2022            NaN  1996-1944  NaN   
24148           2023            NaN  1435-4934  NaN   
24149           2023            NaN  1354-0661  NaN   

                                  doi  projectID projectAcronym  \
0           10.3390/molecules28083419  101040480        LACRIDO   
1           10.1016/j.jms.2023.111824  101040480        LACRIDO   
2                   10.1063/5.02133

### Missin values
Here we inspect the missing data in this file, and outline how we are goiing to treat these missing data points

In [20]:
# look for missing values
publications_missing = data_publications.isnull()

# check which columns are missing data
for key in publications_missing.keys():
    missing = publications_missing[publications_missing[key] == True]
    if len(missing.id) > 0:
        print(f'For key {key}:\n     {len(missing.id)} elements are missing.')

For key authors:
     75 elements are missing.
For key journalTitle:
     2099 elements are missing.
For key journalNumber:
     13622 elements are missing.
For key publishedPages:
     24142 elements are missing.
For key issn:
     7004 elements are missing.
For key isbn:
     23219 elements are missing.
For key doi:
     2293 elements are missing.


There is quite some missing data in this file. Let us go through each line individually.
- authors:
    - This sucks. Would have been very nice to decompose author strings into single authors and make the connections
    - How to treat this: look into the article title string to check whether this one contains more author infromation
- journalTitle:
    - chack in the publication title. Sometimes there one has just copy-pasted the whole article reference
- journalNumber:
    - Not the most relevant parameter in my opinion. Just make all NaN zeros
- publishedYear:
    - Manually look this up
- publishedPages:
    - Not the most relevant parameter in my opinion. Just make all NaN zeros
- issn:
    - Not the most relevant parameter in my opinion. Just make all NaN zeros
- isbn:
    - Not the most relevant parameter in my opinion. Just make all NaN zeros
- doi:
    - Fuck this, just pass about:blank as url. 
- rcn:
    - Manually adjust this one. 
        - Update: this entry was missing an entry for authors, all following field shifted 1 column to the left. Manually fixed this one. 



In [21]:
data_publications.keys()

Index(['id', 'title', 'isPublishedAs', 'authors', 'journalTitle',
       'journalNumber', 'publishedYear', 'publishedPages', 'issn', 'isbn',
       'doi', 'projectID', 'projectAcronym', 'collection', 'contentUpdateDate',
       'rcn'],
      dtype='object')

In [22]:
# check missing rcn. 
data_publications[publications_missing.rcn == True]

Empty DataFrame
Columns: [id, title, isPublishedAs, authors, journalTitle, journalNumber, publishedYear, publishedPages, issn, isbn, doi, projectID, projectAcronym, collection, contentUpdateDate, rcn]
Index: []

In [23]:
# fill some gaps in the data structure
data_publications['isbn'] = data_publications['isbn'].fillna('0000-0000')
data_publications['issn'] = data_publications['issn'].fillna('0000-0000')
data_publications['publishedPages'] = data_publications['publishedPages'].fillna(0)
data_publications['doi'] = data_publications['doi'].fillna('about:blank')
data_publications['journalTitle'] = data_publications['journalTitle'].fillna('Miscalleneous')
data_publications['journalNumber'] = data_publications['journalNumber'].fillna(0)
data_publications['authors'] = data_publications['authors'].fillna('sine nome')


In [24]:
# check data_publications again
publications_missing = data_publications.isnull()

# check which columns are missing data
for key in publications_missing.keys():
    missing = publications_missing[publications_missing[key] == True]
    if len(missing.id) > 0:
        print(f'For key {key}:\n     {len(missing.id)} elements are missing.')

Now there are no empty entries left. We store the completed dataset in the interim folder

In [25]:
# Save to intermediate
data_publications.to_csv(f'{interim_dir}/projectPublications_interim.csv', sep=';')

## Inspect CORDIS-HORIZON projects data files
This is the folder containing some more datasets on the different projects.

In [26]:
# define file paths
CORDIS_framework_docs_dir = f'{raw_dir}/cordis-HORIZONprojects-csv'

SciVoc_path = f'{CORDIS_framework_docs_dir}/euroSciVoc.csv'
legalBasis_path = f'{CORDIS_framework_docs_dir}/legalBasis.csv'
organization_path = f'{CORDIS_framework_docs_dir}/organization.csv'
project_path = f'{CORDIS_framework_docs_dir}/project.csv'
topics_path = f'{CORDIS_framework_docs_dir}/topics.csv'
webItems_path = f'{CORDIS_framework_docs_dir}/webItem.csv'
webLink_path = f'{CORDIS_framework_docs_dir}/webLink.csv'

In [27]:
# Import some informative files

# load datasets
read_csv_options = {
    "delimiter": ";",
    "quotechar": '"',
    "escapechar": "\\",
    'doublequote': False,
    # "on_bad_lines": "skip",   # we skip lines that do not import properly for now
    "engine": "python"  # 'python' engine handles complex parsing better
}


sci_voc_df = pd.read_csv(SciVoc_path, **read_csv_options)
legal_basis_df = pd.read_csv(legalBasis_path, **read_csv_options)
organization_df = pd.read_csv(organization_path, delimiter=';')
topics_df = pd.read_csv(topics_path, **read_csv_options)
web_items_df = pd.read_csv(webItems_path, **read_csv_options)
web_link_df = pd.read_csv(webLink_path, **read_csv_options)


In [28]:
read_csv_options['on_bad_lines'] = 'skip'
try:
    project_df = pd.read_csv(project_path, **read_csv_options)
    print(len(project_df.id))
except pd.errors.ParserError as e:
    print("Parsing error:", e)

15737


There are 15863 - 15737 = 126 field which could not be read. Let us inspect those lines further. 

In [117]:
project_df

"id"        "acronym"  "status"  \
0      "101159220"      "PvSeroRDT"  "SIGNED"   
1      "101096150"       "BIOBoost"  "SIGNED"   
2      "101093997"  "GlycanTrigger"  "SIGNED"   
3      "101126531"   "CHIKVAX_CHIM"  "SIGNED"   
4      "101113979"      "The Oater"  "CLOSED"   
...            ...              ...       ...   
12674  "101146604"    "EU TalentON"  "SIGNED"   
12675  "101052410"      "EUCYS2022"  "CLOSED"   
12676  "101124648"   "RESAVER_2023"  "SIGNED"   
12677  "101172981"      "EUCYS2024"  "SIGNED"   
12678  "101131799"       "CO-VALUE"  "SIGNED"   

                                                 "title"   "startDate"  \
0      "A point-of-care serological rapid diagnostic ...  "2025-02-01"   
1      "Boosting innovation agencies for bioeconomy v...  "2023-02-01"   
2      "GLYCANS AS MASTER TRIGGERS OF HEALTH TO INTES...  "2023-01-01"   
3      "Late-stage clinical development of Chikunguny...  "2023-06-01"   
4      "The Oater develops a compact machine for hype...  "2023-07-01"   
...                                                  ...           ...   
12674  "EU TalentON Katowice 2024 - European Union Co...  "2023-12-01"   
12675                                 "EUCYS Leiden2022"  "2021-09-01"   
12676  "Support to Retirement Savings Vehicle for Eur...  "2023-09-01"   
12677  "European Union Contest for Young Scientists (...  "2024-02-01"   
12678  "Citizen-Oriented Valorisation for Advancement...  "2024-01-01"   

          "endDate"   "totalCost" "ecMaxContribution"   "legalBasis"  \
0      "2030-01-31"  "4062396,23"        "4062396,23"  "HORIZON.2.1"   
1      "2025-01-31"           "0"            "500000"  "HORIZON.3.2"   
2      "2028-12-31"     "6771571"           "6771571"  "HORIZON.2.1"   
3      "2028-11-30"   "100000000"          "70000000"  "HORIZON.2.1"   
4      "2023-12-31"           "0"             "75000"  "HORIZON.3.2"   
...             ...           ...                 ...            ...   
12674  "2024-11-30"    "499691,5"          "499691,5"  "HORIZON.4.2"   
12675  "2023-02-28"     "2000000"           "2000000"  "HORIZON.4.2"   
12676  "2026-08-31"  "2499638,25"        "2499638,25"  "HORIZON.4.2"   
12677  "2025-02-28"      "999500"            "999500"  "HORIZON.4.2"   
12678  "2026-12-31"     "1046625"           "1046625"  "HORIZON.4.2"   

                                          "topics" "ecSignatureDate"  \
0      "HORIZON-JU-GH-EDCTP3-2023-02-02-two-stage"      "2024-12-09"   
1                 "HORIZON-EIE-2022-CONNECT-01-01"      "2022-11-25"   
2               "HORIZON-HLTH-2022-STAYHLTH-02-01"      "2022-12-05"   
3               "HORIZON-HLTH-2022-CEPI-15-01-IBA"      "2023-06-15"   
4                 "HORIZON-EIE-2022-SCALEUP-02-02"      "2023-06-05"   
...                                            ...               ...   
12674            "HORIZON-WIDERA-2023-EU-CECR-IBA"      "2023-11-30"   
12675              "HORIZON-WIDERA-2021-EUCYS-IBA"      "2022-02-07"   
12676            "HORIZON-WIDERA-2023-RESAVER-IBA"      "2023-07-10"   
12677              "HORIZON-WIDERA-2024-EUCYS-IBA"      "2024-04-15"   
12678              "HORIZON-WIDERA-2023-ERA-01-03"      "2023-10-05"   

      "frameworkProgramme"                              "masterCall"  \
0                "HORIZON"  "HORIZON-JU-GH-EDCTP3-2023-02-two-stage"   
1                "HORIZON"             "HORIZON-EIE-2022-CONNECT-01"   
2                "HORIZON"           "HORIZON-HLTH-2022-STAYHLTH-02"   
3                "HORIZON"           "HORIZON-HLTH-2022-CEPI-15-IBA"   
4                "HORIZON"             "HORIZON-EIE-2022-SCALEUP-02"   
...                    ...                                       ...   
12674            "HORIZON"         "HORIZON-WIDERA-2023-EU-CECR-IBA"   
12675            "HORIZON"           "HORIZON-WIDERA-2021-EUCYS-IBA"   
12676            "HORIZON"         "HORIZON-WIDERA-2023-RESAVER-IBA"   
12677            "HORIZON"           "HORIZON-WIDERA-2024-EUCYS-IBA"   
12678         

## Construct functions to access cleaned data

We now define some functions that allow easy access to all aspects of different projects. 


- Merge datasets into one object
- Standardize column names => they are compatible
- Create function that allow access to project-specific data:
    - function argument: project name / acronym / identifier
    - function output: data class with project information as attributes
    - Or: approach this from a class init perspective

Find some way to pass load datasets
apply class on this, without having to load the full dataset each time we initialize the class


In [ ]:
# class to load datasets

class CORDIS_data():
    def __init__(self):
        '''
        Initialize class: load data from the CSV files

        set some global variables that we 

        '''

        self.data_report = pd.read_csv(f'{interim_dir}/projectdeliverables_interim.csv', delimiter=';')
        self.data_deliverables = pd.read_csv(f'{interim_dir}/projectdeliverables_interim.csv', delimiter=';')

        pass

    
    def list_of_acronyms(self):
        '''
        This function prints out a dataframe 
        '''
        pass

class Project_data(CORDIS_data):
    ''' 
    This class inherits all the attributes from the class CORDIS_data, including the loaded datasets. 
    Intended use case: 
        set some new variables related to a project:
            - number of deliverables
            - deliverable_types
            - number of publications
            - publication types
            - ... (whatever might be useful later on in the project)
    '''
    
    def project(self, id=None, acronym=None):
        '''
        class to access all information related to a certain project:
        create some a

        '''
        pass


In [ ]:
# store post-processed and feature enriched datasets here

data_full.to_csv(f'{processed_dir}/CORDIS_enriched.csv', sep=';')